File: fze-commons/src/main/java/net/fze/commons/std/api/client/ApiClient.java
package net.fze.commons.std.api.client;

import net.fze.commons.http.HttpClient;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import net.fze.commons.Types;
import net.fze.commons.TypesConv;
import net.fze.commons.std.api.ApiUtils;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

/**
 * API客户端
 */
public class ApiClient {
    private String _apiUrl;
    private String _key;
    private String _secret;
    private String _version;
    private String _signType;
    private int _timeout;
    private Map<String, String> extraParams = new HashMap<>();

    public ApiClient(String apiUrl, String key, String secret, String version, String signType, int timeout) {
        this._apiUrl = apiUrl;
        this._key = key;
        this._secret = secret;
        this._version = version;
        this._signType = signType;
        this._timeout = timeout;
    }

    public String getApiUrl() {
        return this._apiUrl;
    }

    public String getKey() {
        return this._key;
    }

    public String getSecret() {
        return this._secret;
    }

    public String getSignType() {
        return this._signType;
    }

    /**
     * 添加额外的参数
     *
     * @param key   参数名
     * @param value 值
     */
    public void addParam(String key, String value) {
        this.extraParams.put(key, value);
    }

    private Exception except(String ret) {
        ret = ret.substring(1);
        int i = ret.indexOf('#');
        int code = TypesConv.toInt(ret.substring(0, i));
        String message = ret.substring(i + 1);
        return new Exception(String.format("#%d:%s", code, message));
    }

    /**
     * 请求接口，返回原始内容
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应字节
     */
    public byte[] request(String apiName, Map<String, String> params) throws Exception {
        Map<String, String> data = Types.cloneMap(params);
        Types.copyMap(this.extraParams, data);
        data.put("api", apiName);
        data.put("key", this._key);
        data.put("version", this._version);
        data.put("sign_type", this._signType);
        data.put("sign", ApiUtils.Sign(this._signType, data, this._secret));
        String query = HttpClient.toQuery(data);
        return HttpClient.request(this._apiUrl, "POST", query.getBytes(), this._timeout);
    }

    /**
     * 调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    public String call(String apiName, Map<String, String> params) throws Exception {
        String ret = new String(this.request(apiName, params));
        if (ret.startsWith("#")) {
            throw this.except(ret);
        }
        return ret;
    }


    /**
     * 支持泛型的调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    public <T> T callByClass(String apiName, Map<String, String> params, Class<?> classOfT, Class<?>... classOfArgs) throws Exception {
        String ret = new String(this.request(apiName, params));
        if (ret.length() == 0) return null;
        if (ret.startsWith("#")) {
            throw this.except(ret);
        }
        Type gt;
        if (classOfArgs.length == 0) {
            gt = TypeToken.get(classOfT).getType();
        } else {
            gt = TypeToken.getParameterized(classOfT, classOfArgs).getType();
        }
        return new Gson().fromJson(ret, gt);
    }

    /**
     * 支持泛型的调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    public <T> T callByType(String apiName, Map<String, String> params, Type gt) throws Exception {
        String ret = new String(this.request(apiName, params));
        if (ret.length() == 0) return null;
        if (ret.startsWith("#")) {
            throw this.except(ret);
        }
        return new Gson().fromJson(ret, gt);
    }


}

File: fze-commons/src/main/java/net/fze/api/client/ApiClient.kt
package net.fze.api.client

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import net.fze.commons.std.Types
import net.fze.commons.std.TypesConv
import net.fze.commons.std.http.HttpClient
import java.lang.reflect.Type

/**
 * API客户端
 */
class ApiClient {
    private val expires: Int
    private var lastTokenUnix: Int = 0
    private var accessToken: String = ""
    private var accessFunc: IAccessToken
    private var apiUrl = ""
    private var key = ""
    private var secret = ""
    private var _timeout = 0
    private val extraParams: MutableMap<String, String> = HashMap()

    constructor(apiUrl: String, key: String, secret: String,a:IAccessToken,tokenExpires:Int,timeout: Int){
        this.apiUrl = apiUrl
        this.key = key
        this.secret = secret
        this._timeout = timeout
        this.accessFunc = a
        this.expires = tokenExpires
    }

    /**
     * 添加额外的参数
     *
     * @param key   参数名
     * @param value 值
     */
    fun addParam(key: String, value: String) {
        extraParams[key] = value
    }

    private fun except(ret: String): Exception {
        var ret = ret
        ret = ret.substring(1)
        val i = ret.indexOf('#')
        val code = TypesConv.toInt(ret.substring(0, i))
        val message = ret.substring(i + 1)
        return Exception(String.format("#%d:%s", code, message))
    }

    /**
     * 请求接口，返回原始内容
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应字节
     */
    @Throws(Exception::class)
    fun request(apiName: String, params: Map<String, String>?): ByteArray {
        val data = Types.cloneMap(params)
        Types.copyMap(extraParams, data)
        data["user_key"] = key
        val query = HttpClient.toQuery(data)
        var now = Types.time.unix()
        if(now-this.lastTokenUnix > this.expires){
            this.accessToken = this.accessFunc.get(this.key, this.secret)
            this.lastTokenUnix = now
        }
        var headers = mutableMapOf<String,String>()
        headers["Authorization"] = this.accessToken
        return HttpClient.request(this.concat(apiName), "POST",
                query.toByteArray(),
                headers, _timeout)
    }

    private fun concat(apiName: String): String {
        var path = apiName
        if(path.length > 0 && path[0]  !='/'){
            path = "/"+path
        }
        return this.apiUrl + path
    }

    /**
     * 调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    @Throws(Exception::class)
    fun call(apiName: String, params: Map<String, String>?): String {
        val ret = String(request(apiName, params))
        if (ret.startsWith("#")) {
            throw except(ret)
        }
        return ret
    }

    /**
     * 支持泛型的调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    @Throws(Exception::class)
    fun <T> callByClass(apiName: String, params: Map<String, String>?, classOfT: Class<*>?, vararg classOfArgs: Class<*>?): T? {
        val ret = String(request(apiName, params))
        if (ret.length == 0) return null
        if (ret.startsWith("#")) {
            throw except(ret)
        }
        val gt: Type
        gt = if (classOfArgs.size == 0) {
            TypeToken.get(classOfT).type
        } else {
            TypeToken.getParameterized(classOfT, *classOfArgs).type
        }
        return Gson().fromJson(ret, gt)
    }

    /**
     * 支持泛型的调用接口
     *
     * @param apiName 接口名称
     * @param params  参数
     * @return 响应
     */
    @Throws(Exception::class)
    fun <T> callByType(apiName: String, params: Map<String, String>?, gt: Type?): T? {
        val ret = String(request(apiName, params))
        if (ret.length == 0) return null
        if (ret.startsWith("#")) {
            throw except(ret)
        }
        return Gson().fromJson(ret, gt)
    }
}